src/SimpleCLIUserInterface.cpp

```diff
@@ -0,0 +1,352 @@
+// Copyright (C) Sven Rosvall (sven@rosvall.ie)
+// This file is part of VLCB-Arduino project on https://github.com/SvenRosvall/VLCB-Arduino
+// Licensed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
+// The full licence can be found at: http://creativecommons.org/licenses/by-nc-sa/4.0/
+
+#include "SimpleCLIUserInterface.h"
+#include "Controller.h"
+#include <Streaming.h>
+#include <SimpleCLI.h>
+
+extern void printConfig();
+
+
+namespace VLCB
+{
+
+// Create CLI Object
+SimpleCLI simpleCli(CLI_COMMAND_QUEUE_SIZE); // Default of 10 commands
+
+Command cmdHelp;
+Command cmdHelpWithArg;
+Command cmdSerialInterface;
+
+//////////////////////////////////////////////////////////////////
+// See Modern C++ Design, Andrei Alexandrescu, 2001(!) page 116
+// First declare a typedef for a pointer to what is needed.
+// Note the argument type.
+typedef void (SimpleCLIUserInterface::* TpMemFun)(cmd *);
+// Then create an object which implements tha callback
+TpMemFun pSerialCallback = &SimpleCLIUserInterface::serialCallback;
+// An instance of class SimpleUserInterace is also needed.
+// Where is it? Declared in the top level code.
+// This is the actual call to use the callback
+//     (simpleCLIUserInterface.*pSerialCallback)(c);
+// It is used in local_templateCallback(cmd *c) - see below.
+/////////////////////////////////////////////////////////////
+
+
+
+SimpleCLIUserInterface::SimpleCLIUserInterface(Transport *transport)
+  : transport(transport)
+{
+}
+
+void SimpleCLIUserInterface::setController(Controller *ctrl)
+{
+  this->controller = ctrl;
+  this->modconfig = ctrl->getModuleConfig();
+}
+
+void SimpleCLIUserInterface::process(const Action *action)
+{
+  handleAction(action);
+  
+  //This replaces processSerialInput(); 
+  getHelp();
+}
+
+// This will hold a local copy of the first char of the input.
+char local_serialChar = '0'; 
+
+void SimpleCLIUserInterface::serialCallback(cmd *c) {
+    Command cmd(c); // Create wrapper object
+    Serial.print("called as ");
+	// Copy the char into the instance.
+	serialChar = local_serialChar;
+    Serial.print(serialChar);
+    Serial.print(" : ");
+    Serial.println(cmd.toString());
+	processSerialInputImpl(serialChar); 
+}
+
+void local_serialCallback(cmd *c) {
+    Command cmd(c); // Create wrapper object
```

Comment:
Try to avoid using a variable name with the same name as a type. This is confusing.
It is bad that the SimpleCLI creates types with all lower case letters.

Reply:
I have been following the examples which came with the library. We have already encountered the problem with "cli", In this part of the code I was attempting to get from a callback outside the main class to one inside it. I had decided that I did not want to make any changes to the third party library SimpleCLI so I had to figure out a solution. See some comments in the example code.